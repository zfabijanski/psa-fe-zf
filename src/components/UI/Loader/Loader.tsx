import React, { FC } from "react";
import styled, { createGlobalStyle } from "styled-components/macro";
import { FocusOn } from "react-focus-on";
import LoaderSvg from "../../../assets/loader.svg";

const HideBodyOverflow = createGlobalStyle`
  body {
    overflow: hidden;
  }
`;

type LoaderProps = React.HTMLAttributes<HTMLDivElement> & {};

export const Loader: FC<LoaderProps> = (props) => (
  <FocusOn>
    <HideBodyOverflow />
    <Mask {...props}>
      <FakeButton />
      <Spinner src={LoaderSvg} />
    </Mask>
  </FocusOn>
);

const Mask = styled.div`
  position: fixed;
  z-index: 101;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;

  &:before {
    content: "";
    z-index: -1;
    opacity: 0.8;
    background-color: ${({ theme }) => theme.newColors.gray20};
    z-index: 100;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
`;

const FakeButton = styled.button`
  // for react-focus-on
  position: absolute;
  right: 100%;
`;

const Spinner = styled.img`
  width: 80px;
  z-index: 102;
  position: fixed;
  left: calc(50% - 40px);
  top: calc(50% - 40px);
  animation: rotate cubic-bezier(0.1, 0.3, 0.8, 0.1) 1s infinite;

  @keyframes rotate {
    100% {
      transform: rotate(360deg);
    }
  }
`;
