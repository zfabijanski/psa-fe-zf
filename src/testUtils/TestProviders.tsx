import { render, RenderOptions } from "@testing-library/react";
import { FC, PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { Normalize } from "styled-normalize";
import GlobalStyles from "theme/globalStyles";
import { theme } from "theme/theme";
import rootStore, { configureStore, rootReducer as reducer } from "AppStore";
import { BrowserRouter } from "react-router-dom";
import { Modal } from "components/UI/Modal";
import { IntlProvider } from "react-intl";
import { plMessages } from "translations/pl";
import I18nProvider from "I18nProvider";

interface TestProvidersProps {
  store?: typeof rootStore;
  renderModal?: boolean;
  realIntlProvider?: boolean;
}

export const addTestProviders =
  ({
    store = configureStore({ reducer }),
    renderModal = true,
    // @TODO: update tests to use realIntlProvider everywhere
    realIntlProvider = false,
  }: TestProvidersProps): FC<PropsWithChildren<{}>> =>
  ({ children }) => {
    const TranslationsProvider = realIntlProvider ? IntlProvider : I18nProvider;

    return (
      <Provider store={store}>
        <BrowserRouter>
          <ThemeProvider theme={theme}>
            <TranslationsProvider locale="pl" messages={plMessages}>
              <Normalize />
              <GlobalStyles />
              {children}
              {renderModal && <Modal />}
            </TranslationsProvider>
          </ThemeProvider>
        </BrowserRouter>
      </Provider>
    );
  };

const customRender = (
  ui: ReactElement,
  options: Omit<RenderOptions, "options"> = {},
  testProviderOptions: TestProvidersProps = {}
) => render(ui, { wrapper: addTestProviders(testProviderOptions), ...options });

export * from "@testing-library/react";

// override render method
export { customRender as render };
