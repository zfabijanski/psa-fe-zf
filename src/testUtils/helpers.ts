export const sleep = (ms: number = 100): Promise<unknown> =>
  new Promise((resolve) => setTimeout(resolve, ms));
