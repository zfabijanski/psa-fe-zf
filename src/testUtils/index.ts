export * from "./TestProviders";
export * from "./mockCreators";
export * from "./helpers";
