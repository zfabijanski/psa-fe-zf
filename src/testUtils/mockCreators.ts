import {
  ICalculatorCoversConfig,
  IProductCovers,
} from "features/Calculator/types";
import {
  DocumentState,
  IDirectory,
  IDirectoryMetadata,
  IDocumentMetadata,
} from "features/Library/types";
import { IMeeting } from "features/Meeting/History/types";
import { IIllustration, ProductGuid } from "features/Products/types";
import { IModalData, ModalTypes } from "models/Modal";
import {
  Frequency,
  IFundRiskProfileConfig,
  IProductConfig,
  IProductFund,
  IProductGuarantee,
  InvestmentProfile,
  Role,
} from "models/common";
import {
  CalculationStatus,
  Direction,
  ICalculationDetailsDto,
  ICover,
  IInsured,
  YesNo,
} from "models/calculator";

export const createCalculationDetailsDto = (
  overrides?: Partial<ICalculationDetailsDto>
): ICalculationDetailsDto => ({
  calculation_id: 1,
  working_date: "2020-01-01",
  premiums_total: 100,
  guaranteed_death_benefit: 100,
  guaranteed_maturity_benefit: 100,
  projected_maturity_value_egr_m: 100,
  product_categories: [],
  uuid: "uuid",
  version: "version",
  source_system: "source_system",
  product_guid: 1,
  insureds_list: [],
  covers_list: [],
  funds_list: [],
  ...overrides,
});

export const createInsured = (overrides?: Partial<IInsured>): IInsured => ({
  insured_idx: 1,
  insured_uuid: "insured_uuid",
  age_in_months: 1,
  date_of_birth: "2020-01-01",
  role: Role.MAIN,
  ...overrides,
});

export const createCover = (overrides?: Partial<ICover>): ICover => ({
  cover_id: 1,
  cover_idx: 1,
  insured_idx: 1,
  code: "code",
  frequency_code: Frequency.Annually,
  payments: 1,
  direction_of_calculation: Direction.SumAssuredToPremium,
  investment_profile: InvestmentProfile.Balanced,
  indexation: YesNo.Yes,
  sum_assured: 1,
  premium_per_freq: 1,
  duration: 1,
  idx_option: YesNo.Yes,
  rop_option: YesNo.Yes,
  add_sum_assured: 1,
  ending_age: 1,
  exp_mat_ben: 1,
  ...overrides,
});

export const createCalculatorCoversConfig = (
  overrides?: Partial<ICalculatorCoversConfig>
): ICalculatorCoversConfig => ({
  coverName: "coverName",
  isChild: false,
  isHospital: false,
  ...overrides,
});

export const createCalculatorCoversConfigMap = (
  overrides?: Partial<Record<string, ICalculatorCoversConfig>>
): Record<string, ICalculatorCoversConfig> => ({
  coverName1: createCalculatorCoversConfig({ coverName: "coverName1" }),
  coverName2: createCalculatorCoversConfig({ coverName: "coverName2" }),
  ...overrides,
});

export const createProductCovers = (
  overrides?: Partial<IProductCovers>
): IProductCovers => ({
  mainCover: "mainCover",
  additionalCovers: ["additionalCover1", "additionalCover2"],
  wopCovers: ["wopCover1", "wopCover2"],
  childCovers: ["childCover1", "childCover2"],
  ...overrides,
});

export const createProductCoversMap = (
  overrides?: Partial<Record<string, IProductCovers>>
): Record<string, IProductCovers> => ({
  mainCover1: createProductCovers({ mainCover: "mainCover1" }),
  mainCover2: createProductCovers({ mainCover: "mainCover2" }),
  ...overrides,
});

export const createProductConfig = (
  overrides?: Partial<IProductConfig>
): IProductConfig => ({
  product_guid: 1,
  product_name: "product_name",
  product_categories: [],
  product_freqs: [],
  product_covers: [],
  product_funds: [],
  ...overrides,
});

export const createProductConfigMap = (
  overrides?: Partial<Record<string, IProductConfig>>
): Record<string, IProductConfig> => ({
  1: createProductConfig({ product_name: "product_name1" }),
  2: createProductConfig({ product_name: "product_name2" }),
  ...overrides,
});

export const createFundRiskProfileConfig = (
  overrides?: Partial<IFundRiskProfileConfig>
): IFundRiskProfileConfig => ({
  code: "code",
  name_pl: "name_pl",
  name_en: "name_en",
  risk_level: 1,
  status_code: "status_code",
  notes: "notes",
  ...overrides,
});

export const createProductGuarantee = (
  overrides?: Partial<IProductGuarantee>
): IProductGuarantee => ({
  guarantee_code: "guarantee_code",
  guarantee_name: "guarantee_name",
  guarantee_min_value: "1",
  guarantee_duration_months: "4",
  guarantee_percentages: [{ guarantee_percent: 10 }],
  ...overrides,
});

export const createProductFund = (
  overrides?: Partial<IProductFund>
): IProductFund => ({
  cover_code: "cover_code",
  fund_code: "fund_code",
  fund_name: "fund_name",
  fund_risk_profile: "code",
  fund_order: 1,
  guarantees: [createProductGuarantee()],
  ...overrides,
});

export const createModalData = (
  overrides: Partial<IModalData> = {}
): IModalData => ({
  modalIllustration: "alert-fill",
  modalTitleTrKey: "modalTitleTrKey",
  modalContentTrKey: "modalContentTrKey",
  modalContent: "modalContent",
  modalButtons: {
    confirm: { textTrKey: "modal.ok" },
  },
  modalType: ModalTypes.default,
  maskClosable: true,
  ...overrides,
});

export const createIllustration = (
  overrides: Partial<IIllustration> = {}
): IIllustration => ({
  idd_id: 1,
  product_guid: ProductGuid.Savings,
  adequacy_id: 1,
  order_id: 1,
  expected_amount: { value: 3 },
  premium_total: { value: 3 },
  duration: { value: 3 },
  premium_main_cover: { value: 3 },
  frequency: { value: 3 },
  sum_of_premiums: { value: 3 },
  sum_assured: { value: 3 },
  investment_profile: { value: 3 },
  pru_calc_calculation_id: 1,
  status: CalculationStatus.Illustrated,
  json_needs_analysis: null,
  ...overrides,
});

export const createDirectory = (
  overrides?: Partial<IDirectory>
): IDirectory => ({
  directoryId: 1,
  directories: [],
  documentsIds: [1, 2, 3],
  ...overrides,
});

export const createDirectoryMetadata = (
  overrides?: Partial<IDirectoryMetadata>
): IDirectoryMetadata => ({
  directoryId: 1,
  name: "name",
  isOpen: false,
  ...overrides,
});

export const createDocumentMetadata = (
  overrides?: Partial<IDocumentMetadata>
): IDocumentMetadata => ({
  documentId: 1,
  guid: "guid",
  name: "name",
  state: DocumentState.NotAttached,
  isStateChanging: false,
  moduleName: "PRU_LIBRARY",
  ...overrides,
});

export const createMeeting = (overrides?: Partial<IMeeting>): IMeeting => ({
  meetingNo: "meetingNo",
  meetingId: 1,
  name: "name",
  age: 18,
  lastAccessDate: new Date("2023-01-01"),
  createDate: new Date("2023-01-01"),
  statusId: 1,
  statusName: "statusName",
  statusDescription: "statusDescription",
  state: "state",
  haveAdequacy: false,
  haveCalculation: false,
  haveIdd: false,
});
