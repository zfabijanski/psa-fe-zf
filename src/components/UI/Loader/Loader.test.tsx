import { Loader } from "./Loader";
import { render } from "testUtils";

describe("Loader", () => {
  it("should render correctly", () => {
    const { container } = render(<Loader />);
    expect(container).toMatchSnapshot();
  });
});
