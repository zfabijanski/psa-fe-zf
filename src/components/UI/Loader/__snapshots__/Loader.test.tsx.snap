// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader should render correctly 1`] = `
.c0 {
  position: fixed;
  z-index: 101;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c0:before {
  content: "";
  z-index: -1;
  opacity: 0.8;
  background-color: #EDEDED;
  z-index: 100;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c1 {
  position: absolute;
  right: 100%;
}

.c2 {
  width: 80px;
  z-index: 102;
  position: fixed;
  left: calc(50% - 40px);
  top: calc(50% - 40px);
  -webkit-animation: rotate cubic-bezier(0.1,0.3,0.8,0.1) 1s infinite;
  animation: rotate cubic-bezier(0.1,0.3,0.8,0.1) 1s infinite;
}

<div>
  <div
    aria-hidden="true"
    data-focus-guard="true"
    data-focus-on-hidden="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    data-focus-lock-disabled="false"
  >
    <div
      class="c0"
    >
      <button
        class="c1"
      />
      <img
        class="c2"
        src="loader.svg"
      />
    </div>
  </div>
  <div
    aria-hidden="true"
    data-focus-guard="true"
    data-focus-on-hidden="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-focus-on-hidden="true"
  />
</div>
`;
